app/com/linkedin/drelephant/tuning/engine/MRExecutionEngine.java

```diff
@@ -0,0 +1,458 @@
+    Double usageVirtualMemory = usageStats.get(UsageCounterSchema.USED_VIRTUAL_MEMORY.ordinal());
+    Double usageHeapMemory = usageStats.get(UsageCounterSchema.USED_HEAP_MEMORY.ordinal());
+    Double memoryMB =
+        applyContainerSizeFormula(constraints.get("CONTAINER_MEMORY"), usagePhysicalMemory, usageVirtualMemory);
+    applyHeapSizeFormula(constraints.get("CONTAINER_HEAP"), usageHeapMemory, memoryMB);
+  }
+
+  @Override
+  public void parameterOptimizerIPSO(List<AppResult> results, JobExecution jobExecution) {
+    Map<String, Map<String, Double>> previousUsedMetrics = extractParameterInformationIPSO(results);
+    List<TuningParameterConstraint> parameterConstraints = TuningParameterConstraint.find.where().
+        eq("job_definition_id", jobExecution.job.id).findList();
+    for (String function : functionTypes) {
+      logger.info(" Optimizing Parameter Space  " + function);
+      if (previousUsedMetrics.get(function)
+          .get(CommonConstantsHeuristic.UtilizedParameterKeys.MAX_PHYSICAL_MEMORY.getValue()) > 0.0) {
+        List<Double> usageStats = extractUsageParameter(function, previousUsedMetrics);
+        Map<String, TuningParameterConstraint> memoryConstraints =
+            filterMemoryConstraint(parameterConstraints, function);
+        memoryParameterIPSO(function, memoryConstraints, usageStats);
+      }
+    }
+  }
+
+  @Override
+  public String parameterGenerationsHBT(List<AppResult> results, List<TuningParameter> tuningParameters) {
+    Map<String, Map<String, Double>> previousUsedMetrics = extractParameterInformationIPSO(results);
+    StringBuffer idParameters = new StringBuffer();
+    for (TuningParameter tuningParameter : tuningParameters) {
+      if (tuningParameter.paramName.equals(
+          CommonConstantsHeuristic.ParameterKeys.MAPPER_MEMORY_HADOOP_CONF.getValue())) {
+        idParameters.append(tuningParameter.id)
+            .append("\t")
+            .append(Math.max(previousUsedMetrics.get("map")
+                .get(CommonConstantsHeuristic.UtilizedParameterKeys.MAX_PHYSICAL_MEMORY.getValue()), previousUsedMetrics
+                .get("map")
+                .get(CommonConstantsHeuristic.UtilizedParameterKeys.MAX_VIRTUAL_MEMORY.getValue()) / 2.1));
+        idParameters.append("\n");
+      }
+      if (tuningParameter.paramName.equals(
+          CommonConstantsHeuristic.ParameterKeys.MAPPER_HEAP_HADOOP_CONF.getValue())) {
+        idParameters.append(tuningParameter.id)
+            .append("\t")
+            .append(previousUsedMetrics.get("map")
+                .get(CommonConstantsHeuristic.UtilizedParameterKeys.MAX_TOTAL_COMMITTED_HEAP_USAGE_MEMORY.getValue()));
+        idParameters.append("\n");
+      }
+    }
+    logger.info(" New Suggested Parameter " + idParameters);
+    return idParameters.toString();
+  }
+
+  @Override
+  public Boolean isParamConstraintViolatedHBT(List<JobSuggestedParamValue> jobSuggestedParamValueList) {
```

Comment:
Nit: isParamConstraintViolated sounds more like violation for single parameter while this will be used to check constraints for multiple parameters. Use plural case instead?

Reply:
Will change in next PR